File: src/repo_service.py
"""
Fetch and track GitHub Actions runs triggered by the signed-in user across monitored repos.
Updates store.repo_runs on the main thread via GLib.idle_add.
"""
import threading

from gi.repository import GLib

from formatters import format_duration
import notify

RUNS_PER_REPO = 20

# run_id -> currently in-progress/queued, tracked across polls so we notify exactly once
# when a run we've seen running transitions to completed (in-memory only, per app session).
_tracked_active_run_ids = set()


def _run_is_in_progress(run):
    return (run.get("status") or "").lower() in ("in_progress", "queued")


def fetch_my_runs(client, owner, repo, login, per_page=RUNS_PER_REPO):
    """Recent workflow runs triggered by `login` in owner/repo, newest first. [] on any error."""
    try:
        return client.get_runs_for_actor(owner, repo, login, per_page=per_page).get("workflow_runs", [])
    except Exception:
        return []


def repo_status_from_runs(runs):
    """Aggregate status dot for a repo: yellow (a run is active) > green/red (last completed) > gray (no runs)."""
    if not runs:
        return "gray"
    if any(_run_is_in_progress(r) for r in runs):
        return "yellow"
    top = runs[0]
    if (top.get("status") or "") == "completed":
        return "green" if top.get("conclusion") == "success" else "red"
    return "gray"


def refresh_repo_runs(on_done=None):
    """Fetch runs for every monitored repo in a background thread; apply to store and notify on main thread."""
    import store

    def worker():
        runs_by_repo, completed_events = _poll_fetch(store.config, store.client)

        def on_main():
            store.repo_runs = runs_by_repo
            for repo_key, run in completed_events:
                _notify_completed(repo_key, run)
            if store.update_tray_menu:
                store.update_tray_menu(force=True)
            if on_done:
                on_done(runs_by_repo)

        GLib.idle_add(on_main)

    threading.Thread(target=worker, daemon=True).start()


# ---------------------------------------------------------------------------
# Private
# ---------------------------------------------------------------------------

def _poll_fetch(config, client):
    """Background-safe: fetch recent runs for every monitored repo. Returns (runs_by_repo, completed_events)."""
    login = (config.get("user") or {}).get("login")
    runs_by_repo = {}
    completed_events = []
    if not login or not client or not client.token:
        return runs_by_repo, completed_events
    for repo_key in config.get("repos") or []:
        if "/" not in repo_key:
            continue
        owner, repo = repo_key.split("/", 1)
        runs = fetch_my_runs(client, owner, repo, login)
        runs_by_repo[repo_key] = runs
        for run in runs:
            run_id = run.get("id")
            if run_id is None:
                continue
            if _run_is_in_progress(run):
                _tracked_active_run_ids.add(run_id)
            elif run_id in _tracked_active_run_ids:
                _tracked_active_run_ids.discard(run_id)
                completed_events.append((repo_key, run))
    return runs_by_repo, completed_events


def _notify_completed(repo_key, run):
    """Show a desktop notification for a run that just finished."""
    success = run.get("conclusion") == "success"
    emoji = "🟢" if success else "🔴"
    result = "succeeded" if success else "failed"
    name = run.get("name") or "Workflow"
    branch = (run.get("head_branch") or "—").strip()
    head = run.get("head_commit") or {}
    raw_commit = (head.get("message") or "").strip().split("\n")[0][:60].strip() or "—"
    commit_msg = raw_commit.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    duration = format_duration(run.get("run_started_at"), run.get("updated_at")) or "—"
    body = f"\r\n<b>{emoji} {name} {result}</b>\r\n      {branch}\r\n      {commit_msg}\r\n      Duration: {duration}"
    notify.notify(title=repo_key, body=body, url=run.get("html_url"))

File: app/main/repoService.ts
/**
 * Fetch and track GitHub Actions runs triggered by the signed-in user across monitored repos.
 * Mirrors src(python)/repo_service.py from the previous GTK build.
 */

import { Notification, shell } from "electron";
import type { AppConfig } from "./config";
import { getWorkflowFilter } from "./config";
import type { GitHubClient, GithubRun } from "./github";

// GitHub's max per_page. Fetching the full page (instead of a small slice) means the
// renderer has enough history in memory to paginate/filter the runs table client-side
// without extra round-trips.
const RUNS_PER_REPO = 100;

/** run id -> currently in-progress/queued, tracked so we notify exactly once per completed run. */
const trackedActiveRunIds = new Set<number>();

function isInProgress(run: GithubRun): boolean {
  const status = (run.status || "").toLowerCase();
  return status === "in_progress" || status === "queued";
}

/** A repo's workflow filter (if any) restricts notifications to the selected workflow names. */
function matchesWorkflowFilter(config: AppConfig, repoKey: string, run: GithubRun): boolean {
  const filter = getWorkflowFilter(config, repoKey);
  if (filter.length === 0) return true;
  return filter.includes(run.name || "Workflow");
}

async function fetchMyRuns(
  client: GitHubClient,
  owner: string,
  repo: string,
  login: string,
  perPage = RUNS_PER_REPO,
): Promise<GithubRun[]> {
  try {
    return await client.getRunsForActor(owner, repo, login, perPage);
  } catch {
    return [];
  }
}

export type RepoStatus = "yellow" | "green" | "red" | "gray";

/** Aggregate status dot for a repo: yellow (a run is active) > green/red (last completed) > gray (no runs). */
export function repoStatusFromRuns(runs: GithubRun[] | undefined): RepoStatus {
  if (!runs || runs.length === 0) return "gray";
  if (runs.some(isInProgress)) return "yellow";
  const top = runs[0];
  if (top.status === "completed") {
    return top.conclusion === "success" ? "green" : "red";
  }
  return "gray";
}

function formatDuration(startedAt: string | null, updatedAt: string | null): string {
  if (!startedAt) return "n/a";
  const start = Date.parse(startedAt);
  if (Number.isNaN(start)) return "n/a";
  const end = updatedAt ? Date.parse(updatedAt) : Date.now();
  if (Number.isNaN(end)) return "n/a";
  const seconds = Math.max(0, Math.round((end - start) / 1000));
  const minutes = Math.floor(seconds / 60);
  const secs = seconds % 60;
  return minutes === 0 ? `${secs}s` : `${minutes}m ${secs}s`;
}

function notifyCompleted(repoKey: string, run: GithubRun): void {
  const success = run.conclusion === "success";
  const emoji = success ? "🟢" : "🔴";
  const result = success ? "succeeded" : "failed";
  const name = run.name || "Workflow";
  const branch = run.head_branch || "—";
  const rawCommit = (run.head_commit?.message || "").trim().split("\n")[0].slice(0, 60).trim() || "—";
  const duration = formatDuration(run.run_started_at, run.updated_at) || "—";

  const notification = new Notification({
    title: repoKey,
    body: `${emoji} ${name} ${result}\n${branch}\n${rawCommit}\nDuration: ${duration}`,
  });
  if (run.html_url) {
    notification.on("click", () => shell.openExternal(run.html_url));
  }
  notification.show();
}

interface PollResult {
  runsByRepo: Record<string, GithubRun[]>;
}

/** Fetch recent runs for every monitored repo; fire notifications for runs that just completed. */
export async function pollAllRepos(
  config: AppConfig,
  client: GitHubClient | null,
  onNotify?: () => void,
): Promise<PollResult> {
  const runsByRepo: Record<string, GithubRun[]> = {};
  const login = config.user?.login;
  if (!login || !client?.token) {
    return { runsByRepo };
  }
  for (const repoKey of config.repos || []) {
    if (!repoKey.includes("/")) continue;
    const [owner, repo] = repoKey.split("/", 2);
    const runs = await fetchMyRuns(client, owner, repo, login);
    runsByRepo[repoKey] = runs;
    for (const run of runs) {
      if (run.id == null) continue;
      if (isInProgress(run)) {
        trackedActiveRunIds.add(run.id);
      } else if (trackedActiveRunIds.has(run.id)) {
        trackedActiveRunIds.delete(run.id);
        if (config.notifyEnabled && matchesWorkflowFilter(config, repoKey, run)) {
          notifyCompleted(repoKey, run);
          onNotify?.();
        }
      }
    }
  }
  return { runsByRepo };
}
